scripts/ssl/extract_features.py

```diff
@@ -0,0 +1,285 @@
+        model = EncDecDenoiseMaskedTokenPredModel.restore_from(model_path)
+    else:
+        logging.info(f"Loading model from pretrained: {model_path}")
+        model = EncDecDenoiseMaskedTokenPredModel.from_pretrained(model_name=model_path)
+    return model
+
+
+class FeatureExtractor(pl.LightningModule):
+    """
+    Wrapper class for extracting features from SSL model
+    """
+
+    def __init__(self, ssl_model: EncDecDenoiseMaskedTokenPredModel, layer: str = "all"):
+        super().__init__()
+        self.preprocessor = ssl_model.preprocessor
+        self.encoder = ssl_model.encoder
+        self.layer_idx_list = None
+        self.sample_rate = ssl_model.cfg.sample_rate
+        if layer == "all":
+            self.layer_idx_list = None
+        elif layer == "last":
+            self.layer_idx_list = [len(self.encoder.layers) - 1]
+        else:
+            try:
+                self.layer_idx_list = [int(l) for l in layer.split(",")]
+            except Exception as e:
+                raise ValueError(f"Invalid layer argument: {layer}. Error: {e}")
+        self.feature_extractor = ConformerMultiLayerFeatureExtractor(
+            self.encoder, aggregator=None, layer_idx_list=self.layer_idx_list
+        )
+
+    def forward(
+        self,
+        input_signal=None,
+        input_signal_length=None,
+        processed_signal=None,
+        processed_signal_length=None,
+    ):
+        """
+        Forward pass to extract features, same input interface as EncDecDenoiseMaskedTokenPredModel.forward
+        """
+        has_input_signal = input_signal is not None and input_signal_length is not None
+        has_processed_signal = processed_signal is not None and processed_signal_length is not None
+        if (has_input_signal ^ has_processed_signal) == False:
+            raise ValueError(
+                f"{self} Arguments ``input_signal`` and ``input_signal_length`` are mutually exclusive "
+                " with ``processed_signal`` and ``processed_signal_len`` arguments."
+            )
+        if not has_processed_signal:
+            processed_signal, processed_signal_length = self.preprocessor(
+                input_signal=input_signal,
+                length=input_signal_length,
+            )
+        encoded, encoded_len = self.feature_extractor(audio_signal=processed_signal, length=processed_signal_length)
+        return encoded, encoded_len
+
+
+def maybe_save_features(output_dir, results, max_cache, manifest):
+    """
+    Check if the cache is full and save features to disk
+    """
+    if len(results) == 0 or max_cache < 0 or len(results) < max_cache:
+        return
+    os.makedirs(output_dir, exist_ok=True)
+    logging.info(f"Saving {len(results)} features to {output_dir}")
+
+    for sample_id, audio_file, features_np in tqdm(results, desc="Saving features", total=len(results)):
+        filename = str(audio_file).replace("/", "_").replace(".", "_")
+        if len(filename) > 256:
```

Comment:
Hmm this shouldnt be silent - how about making it a param of the script ?

Reply:
the 256 limit is from filesystem, so I hard-coded it there